### ISSUE TITLE
KeyExpired middleware is not able to retrieve a session object when session is handled by a custom Go plugin

### ISSUE THREAD

**Issue by Ruto8** [TykTechnologies/tyk#3627]
**Branch/Environment/Version**
- Branch/Version: 3.2.2-rc4

**Describe the bug**
KeyExpired middleware is not able to retrieve a session object when session is handled by a custom Go plugin

**Reproduction steps**

I've prepared this repository here where you'll be able to reproduce the issue:
https://github.com/Ruto8/tyk-gateway-docker

1. Make a request that should be a successful request according to the authentication plugin implementation. Using docker compose from the above repository that would be:
```bash
curl --location --request GET 'localhost:8080/tyk-auth-test/' --header 'Authorization: test' --header 'X-Should-Authorize: true'
```

**Actual behavior**
Response: 
```
{
    "error": "Session state is missing or unset! Please make sure that auth headers are properly applied"
}%                                                                                                                                                                                      
```

**Expected behavior**
I expected the request to go through the chain of middleware and eventually be proxied to httpbin, as specified in the API definition.

**Configuration (tyk config file):**
Available in the repository with reproduction of the bug

**Additional context**
We're currently working through the recent issues about go plugins with Tyk. Here are somewhat related issues 
- https://github.com/TykTechnologies/tyk/issues/3605
- https://github.com/TykTechnologies/tyk/issues/3597
- https://github.com/TykTechnologies/tyk/issues/3573

(Alexis, creator of one of the issues above is actually my team member). 
The issues above are the reason we're using `v3.2.2-rc4` currently. Thanks to that we were able to get past them, but to the point of the issue described above.

I'm also aware that I'm describing an issue with an unstable version now, but it seems like it's the only workaround for us at this point. I wanted you to be aware, and maybe hopefully there's a workaround, or there is an issue with my setup.
Thanks!

**Comment by Ruto8** [comment]
@christtyk @alephnull @buger 
sorry for pinging you all, but this is blocking us from implementing Tyk as a part of our app. 
Is there any workaround for the above problem? Or do you know any version of Tyk that will work for us? (As linked the other issues, we've tried several different older versions and none worked)

Is there anything we can do to help?
Thanks

**Comment by christtyk** [comment]
@buger will be best to weigh in here. 

@matiasinsaurralde may be able to support as it's middleware related as well. 

@Ruto8 assuming you're using the open source version right now? Appreciate you passing on these reproduction steps + will flag this to the team to respond as soon as possible.

**Comment by Ruto8** [comment]
Thank you for the quick reply

> @Ruto8 assuming you're using the open source version right now? 

Correct.

**Comment by Darkweak** [comment]
Might related to https://github.com/TykTechnologies/tyk/issues/3612 @Ruto8

**Comment by emwalker** [comment]
@Ruto8 I'm seeing a similar error when I attempt to load and use a Go auth plugin.  When I do that, the session isn't visible to the downstream `KeyExpired` middleware. When I checked out your fork of the docker repo, I couldn't find a Go plugin.  Was your purpose in linking to the forked repo just to show what the error looks like?

**Comment by andyo-tyk** [comment]
This was resolved in Tyk 4.0.
